test_agent/rust/src/testagent.rs

```diff
@@ -0,0 +1,293 @@
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use async_trait::async_trait;
+use log::error;
+use serde_json::Value;
+use up_rust::{Data, UCode, UListener};
+use up_rust::{UMessage, UStatus, UTransport};
+
+use std::io::{Read, Write};
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use serde::Serialize;
+
+use crate::utils::{convert_json_to_jsonstring, WrapperUMessage, WrapperUUri};
+use crate::{constants, utils, TcpStreamSync, UTransportSocket};
+
+use self::utils::sanitize_input_string;
+
+#[derive(Serialize)]
+pub struct JsonResponseData {
+    data: HashMap<String, String>,
+    action: String,
+    ue: String,
+    test_id: String,
+}
+#[derive(Clone)]
+pub struct SocketTestAgent {
+    utransport: UTransportSocket,
+    clientsocket: Arc<Mutex<TcpStreamSync>>,
+    clientsocket_to_tm: Arc<Mutex<TcpStreamSync>>,
+}
+
+#[async_trait]
+impl UListener for SocketTestAgent {
+    async fn on_receive(&self, msg: UMessage) {
+        dbg!("OnReceive called");
+
+        let data_payload = match &msg.payload.data {
+            Some(data) => {
+                // Now we have access to the Data enum
+                match data {
+                    Data::Reference(reference) => reference.to_string(),
+                    Data::Value(value) => {
+                        let value_str = String::from_utf8_lossy(value);
+                        value_str.to_string()
+                    }
+                    _ => "none".into(),
+                }
+            }
+            None => {
+                dbg!("No data available");
+                "none".into()
+            }
+        };
+
+        let mut value: HashMap<String, String> = HashMap::new();
+        value.insert("value".into(), data_payload);
+        let Ok(value_str) = serde_json::to_string(&value) else {
+            error!("issue in converting to payload");
+            return;
+        };
+
+        let mut payload: HashMap<String, String> = HashMap::new();
+        payload.insert("payload".into(), value_str);
+        let Ok(payload_str) = serde_json::to_string(&payload) else {
+            error!("issue in converting to payload");
+            return;
+        };
+
+        let mut json_message = JsonResponseData {
+            action: constants::RESPONSE_ON_RECEIVE.to_owned(),
+            data: HashMap::new(),
+            ue: "rust".to_string(),
+            test_id: "1".to_string(),
+        };
+
+        json_message.data.insert("data".into(), payload_str);
+
+        //<SocketTestAgent as Clone>::clone(&self)
+        //todo: revisit this and check:Better pattern might be to have the UListener be impled on a different struct
+        //e.g. SocketTestListener that then holds an Arc<Mutex<SocketTestAgent>> to be able to call send_to_tm() on that instead.
+        dbg!("sending received data to tm....");
+        self.clone().send_to_tm(json_message).await;
```

Comment:
IMHO not comfortable merging this till the suggestions I gave for restructuring the code to avoid cloning self are integrated.